controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,116 @@
+package nstemplateset
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+)
+
+type spaceRolesManager struct {
+	*statusManager
+}
+
+// ensure ensures that the space roles for the users exist.
+// Returns `true, nil` if something was changed, `false, nil` if nothing changed, `false, err` if an error occurred
+func (r *spaceRolesManager) ensure(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+	logger = logger.WithValues("nstemplateset_name", nsTmplSet.Name)
+
+	nss, err := fetchNamespacesByOwner(r.Client, nsTmplSet.Name)
+	if err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+			"failed to list namespaces for workspace '%s'", nsTmplSet.Name)
+	}
+	logger.Info("ensuring space roles", "namespace_count", len(nss), "role_count", len(nsTmplSet.Spec.SpaceRoles))
+	changed := false
+	for _, ns := range nss {
+		// space roles previously applied
+		// read annotation to see what was applied last time, so we can compare with the new SpaceRoles and remove all outdated resources (based on their kind/names)
+		lastAppliedSpaceRoles := []toolchainv1alpha1.NSTemplateSetSpaceRole{}
+		if currentSpaceRolesAnnotation, exists := ns.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey]; exists && currentSpaceRolesAnnotation != "" {
+			if err := json.Unmarshal([]byte(currentSpaceRolesAnnotation), &lastAppliedSpaceRoles); err != nil {
+				return false, errors.Wrap(err, "unable to decode current space roles in annotation")
+			}
+		}
+		// compare last-applied vs spec to see if there's anything obsolete to do
+		if reflect.DeepEqual(nsTmplSet.Spec.SpaceRoles, lastAppliedSpaceRoles) {
+			logger.Info("no space role to update", "namespace", ns.Name)
+			continue
+		}
```

Comment:
I guess that the code doesn't set the status if the SpaceRoles are not up-to-date:
```suggestion
		}
		if err := r.setStatusUpdatingIfNotProvisioning(nsTmplSet); err != nil {
			return false, err
		}
```